### ISSUE TITLE
Prevent `:` and `,` from combining with alphanumeric characters to form identifiers.

### ISSUE THREAD

**Issue by lukaszcz** [anoma/juvix#1495]
As discussed during the retreat, we should allow users to write e.g. `(A,B)`, `(x, 1)`, `a:A` with the expected meaning.

One way to achieve this without introducing new keywords is to make `:` and `,` not combine with letters to form identifiers. Though we should let `:` combine with another `:` to form `::`. We could let `:` and `,` combine with succeeding non-alphanumeric characters. We don't want to let them combine with preceding non-alphanumeric characters, because we want the users to be able to write e.g. `(A-,B)` or `a#:A`.

**Comment by lukaszcz** [comment]
Regardless of how we solve this or what we ultimately decide, I want to reiterate my strong objection to parsing `(A, B)` into `(`, `A,`, `B` and `)`.

For the comma, the main reason is that this violates a nearly universal convention which goes beyond programming languages. Nowhere except in the weird world of Agda (plus maybe a few other esoteric examples) is a space required in front of a comma. In fact, commas are normally required by convention _not_ to have spaces in front of them, like in this sentence. A comma is always a delimiter.

**Comment by lukaszcz** [comment]
The convention for `:` is less established, but I would (more weakly) also argue in favour of allowing `a:A`.

**Comment by lukaszcz** [comment]
I would just make the comma not combine with anything, like `;`. Having commas in identifiers seems like an extremely minor benefit for the price of violating a universal lexical convention which goes back centuries, causing confusion and inconvenience to almost everyone.

**Comment by lukaszcz** [comment]
> Nowhere except in the weird world of Agda (plus maybe a few other esoteric examples) is a space required in front of a comma.

Maybe Lisp is one such esoteric example, but the comma is used there for evaluation inside backquotes, not as a pairing operator, as far as I know.

**Comment by janmasrovira** [comment]
I&rsquo;ll focus only on the `,` symbol for now.

Here I would like to distinguish two cases.

1.  Allow `a, b` to mean `a , b`.
2.  Allow `a,b` to mean `a , b`.

I would say the first case is more common than the second one and so it makes
sense to consider the possibility to allow commas in the middle of identifiers
but not at the end.

The one benefit that I see from allowing 1 or 2 is that we will follow the widely
spread convention in programming languages that the comma is a key symbol of the
language used only as a delimiter.

If we do not follow this convention we can have two problems:

1.  The user is confused and does not understand why something like `a, b`
    results in `'a,' symbol not in scope`.
2.  The user understands that you need a space after a `,` but gets annoyed by it
    because they are not used to it.

The first problem has a clear solution path. We should provide a clear error
message when we encounter an out of scope symbol with a `,` in it.

The only solution to the second problem that I see is to change the syntax of
the language.

Now, I think it is safe to assume that most users coming from languages other
than Agda will have problem 1. But clear error messages will solve that. Then,
how many users will have the second problem? I have no idea. It is something
that we be nice to test somehow. Of course, this would need to be tested after
addressing the first issue.

The downsides of treating `,` as a special symbol are:

1.  We introduce some special cases in the lexing, making it more complex.
2.  We disallow some identifiers for the sake of convention. Note that `,` is
    different than `;` because we need `;` to be a special symbol in order to
    properly lex the language. This is not the case with `,`.

So, are the identifiers that we forbid ever useful? I (in Agda) have used commas
in identifiers multiple times, specially in combination with mixfix operators.
A couple of examples:

1.  To express the forcing relation in some modal logic. In latex is usually
    typeset as $Π , w ⊩ ϕ$. Then in Agda we can easily replicate that syntax with
    a mixfix operator `_,_⊩_`. Example usage: `soundness : ∀ {w A Π} → Π ⊢ A → M
       , w ⊩* Π → M , w ⊩ A`
2.  To express a line in a Hilbert style proof. For instance we can have
    define the mixfix operator `_[_]_ByMP_,_` and then write something like
    `[ 3 ] (ϕ ↝ ϕ ↝ ϕ) ↝ ϕ ↝ ϕ   ByMP 0 , 1`.
3.  A common pattern in Agda is to name `A,B` an identifier of type `A × B`.

Admittedly we do not have mixfix operators in Juvix and we have not discussed
yet if we want to have them in the future, so examples 1 and 2 are less relevant.

Personally, I think that the best way to proceed is to improve the error
messages first. Then, we evaluate the amount of frustration that some users may
still have. If we think that such frustration is more relevant than having less
special rules, we update the syntax.

**Comment by lukaszcz** [comment]
> 
> 1. Allow `a, b` to mean `a , b`.
> 2. Allow `a,b` to mean `a , b`.
> 

I insist on 1 since it's a universal convention. I would say now we're unnecessarily forcing people to write `a , b` when they mean `a, b`. I don't insist so much on 2.

> 1. To express the forcing relation in some modal logic. In latex is usually
>    typeset as Π,w⊩ϕ. Then in Agda we can easily replicate that syntax with
>    a mixfix operator `_,_⊩_`. Example usage: `soundness : ∀ {w A Π} → Π ⊢ A → M , w ⊩* Π → M , w ⊩ A`
> 2. To express a line in a Hilbert style proof. For instance we can have
>    define the mixfix operator `_[_]_ByMP_,_` and then write something like
>    `[ 3 ] (ϕ ↝ ϕ ↝ ϕ) ↝ ϕ ↝ ϕ   ByMP 0 , 1`.
> 3. A common pattern in Agda is to name `A,B` an identifier of type `A × B`.

The use cases 1 & 2 do not contradict my idea of `,` not combining with anything to form identifiers. The symbol you mention has "holes" between `,` and other stuff. It would imho be more convenient to be able to write `[3] (ϕ ↝ ϕ ↝ ϕ) ↝ ϕ ↝ ϕ ByMP 0, 1`.

For point 3, it's not the end of the world if you adopt a different naming convention without using `,`. In fact, most people who haven't used Agda wouldn't have come up with such a weird idea.

> Then, we evaluate the amount of frustration that some users may still have. If we think that such frustration is more relevant than having less special rules, we update the syntax.

Without any evaluation I can make a reasonable guess that the amount of frustration will be considerable. You're going against an ingrained convention that people learn to apply automatically since primary school.

> 1. We introduce some special cases in the lexing, making it more complex.

What special rules? You just need to put `,` in `reservedSymbols` in Juvix/Compiler/Parser/Lexer.hs.

> 2. We disallow some identifiers for the sake of convention. Note that `,` is
>    different than `;` because we need `;` to be a special symbol in order to
>    properly lex the language. This is not the case with `,`.
> 

We don't need `;` to be a "special" symbol. We could always force the user to write space around it to be recognised as the keyword `;`, and otherwise make it combine with surrounding non-whitespace characters.

From my perspective, this is just ill-conceived attempt at excessive uniformity and "purity" with close to zero practical benefits. Who _really_ needs commas inside a consecutive identifier text? Or maybe someone needs `;`, `.` or `(` in the middle of an identifier? We could enforce spaces around everything. Just always parse consecutive text between spaces as a single keyword/identifier. Then everything will be really uniform with no special rules whatsoever.

**Comment by janmasrovira** [comment]
> What special rules? You just need to put , in reservedSymbols in Juvix/Compiler/Parser/Lexer.hs.

I don't see how that would suffice (keeping in mind that we want to have `,` be a valid constructor and potentially combine with other non-letter characters). I would say a good idea is that we make a pr with this change. Then we it will be clear what special rules we need.

> We don't need ; to be a "special" symbol. We could always force the user to write space around it to be recognised as the keyword ;, and otherwise make it combine with surrounding non-whitespace characters.

What I meant is that `;` is part of the language whether `,` is not (it is a defined symbol). Thus, it makes more sense tha reserve the character `;`.

> From my perspective, this is just ill-conceived attempt at excessive uniformity and "purity" with close to zero practical benefits. Who really needs commas inside a consecutive identifier text? Or maybe someone needs ;, . or ( in the middle of an identifier? We could enforce spaces around everything. Just always parse consecutive text between spaces as a single keyword/identifier. Then everything will be really uniform with no special rules whatsoever.

I perceive frustration that I did not intend to cause. I just wanted to have some discussion about it to convince myself that it was a good change. I can accept the changes without any further push back.

**Comment by lukaszcz** [comment]
> > What special rules? You just need to put , in reservedSymbols in Juvix/Compiler/Parser/Lexer.hs.
> 
> I don't see how that would suffice (keeping in mind that we want to have `,` be a valid constructor and potentially combine with other non-letter characters). I would say a good idea is that we make a pr with this change. Then we it will be clear what special rules we need.
> 

Yes, it is slightly more complicated.

```haskell
rawIdentifier :: [ParsecS r ()] -> ParsecS r Text
rawIdentifier allKeywords = do
  notFollowedBy (choice allKeywords)
  h <- P.satisfy validFirstChar
  t <- P.takeWhileP Nothing validTailChar
  return (Text.cons h t)

validTailChar :: Char -> Bool
validTailChar c =
  isAlphaNum c || validFirstChar c || (c `elem` delimiterSymbols)

delimiterSymbols :: [Char]
delimiterSymbols = ","

reservedSymbols :: [Char]
reservedSymbols = "\";(){}[].≔λ\\"

validFirstChar :: Char -> Bool
validFirstChar c = not (isNumber c || isSpace c || (c `elem` reservedSymbols) || (c `elem` delimiterSymbols))
```

> > From my perspective, this is just ill-conceived attempt at excessive uniformity and "purity" with close to zero practical benefits. Who really needs commas inside a consecutive identifier text? Or maybe someone needs ;, . or ( in the middle of an identifier? We could enforce spaces around everything. Just always parse consecutive text between spaces as a single keyword/identifier. Then everything will be really uniform with no special rules whatsoever.
> 
> I perceive frustration that I did not intend to cause. I just wanted to have some discussion about it to convince myself that it was a good change. I can accept the changes without any further push back.

Sorry, don't ever take this personally. I'm sometimes a bit too direct / "combative" when I see I fail to get my point across, but it's mostly frustration at myself not being able to communicate it well enough.

And my point is that having any syntax at all is largely about catering to commonly agreed-upon conventions and preconceptions of the users. Otherwise why just not use S-expressions? Ignoring any "cultural" aspects, they are clearly superior in terms of language uniformity, ease of implementation, ease of manipulation, etc.

Why have infix operators? They complicate the language, are not strictly necessary, and one can get used to Lisp syntax. But that's not people are familiar with, which causes inconvenience.

A general principle in the design of anything, not just programming languages, is that the design needs to adapt to the user, not the other way round.

What I'm proposing is just the commonly agreed-upon default. I believe any departure from such a default should be extremely well-justified, in terms of tangible benefits to the users, not just some abstract notions of "niceness" or how well it fits with the mental categorizations the developers might have. I don't see such a justification.

**Comment by lukaszcz** [comment]
How ingrained the convention about `,` is, is testified by the fact that some Juvix team members themselves tend to use it when they can, e.g., writing `), ` in SimpleFungibleToken, or line 83 of tests/positive/MiniC/Polymorphism/Input.juvix.

**Comment by janmasrovira** [comment]
@lukaszcz should we close this since we merged #1525?

**Comment by lukaszcz** [comment]
Okay, let's close this.
